src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c

```diff
@@ -0,0 +1,915 @@
+                        spaData.fd,
+                        screen->captureDataReady
+    );
+
+    data->screenProps->captureData = cropTo(
+            spaData,
+            data->rawFormat,
+            screen->captureArea.x, screen->captureArea.y,
+            screen->captureArea.width, screen->captureArea.height
+    );
+
+    screen->captureDataReady = TRUE;
+
+    DEBUG_SCREEN_PREFIX(screen, "data ready\n", NULL);
+    fp_pw_stream_queue_buffer(data->stream, pwBuffer);
+}
+
+static void onStreamStateChanged(
+        void *userdata,
+        enum pw_stream_state old,
+        enum pw_stream_state state,
+        const char *error
+) {
+    struct PwStreamData *data = userdata;
+    DEBUG_SCREEN_PREFIX(data->screenProps, "state %i (%s) -> %i (%s) err %s\n",
+                     old, fp_pw_stream_state_as_string(old),
+                     state, fp_pw_stream_state_as_string(state),
+                     error);
+}
+
+static const struct pw_stream_events streamEvents = {
+        PW_VERSION_STREAM_EVENTS,
+        .param_changed = onStreamParamChanged,
+        .process = onStreamProcess,
+        .state_changed = onStreamStateChanged,
+};
+
+
+static bool startStream(
+        struct pw_stream *stream,
+        uint32_t node
+) {
+    char buffer[1024];
+    struct spa_pod_builder builder =
+            SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+    const struct spa_pod *param;
+
+
+    param = spa_pod_builder_add_object(
+            &builder,
+            SPA_TYPE_OBJECT_Format,
+            SPA_PARAM_EnumFormat,
+            SPA_FORMAT_mediaType,
+            SPA_POD_Id(SPA_MEDIA_TYPE_video),
+            SPA_FORMAT_mediaSubtype,
+            SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw),
+            SPA_FORMAT_VIDEO_format,
+            SPA_POD_CHOICE_ENUM_Id(
+                    2,
+                    SPA_VIDEO_FORMAT_RGBx,
+                    SPA_VIDEO_FORMAT_BGRx
+            ),
+            SPA_FORMAT_VIDEO_size,
+            SPA_POD_CHOICE_RANGE_Rectangle(
+                    &SPA_RECTANGLE(320, 240),
+                    &SPA_RECTANGLE(1, 1),
+                    &SPA_RECTANGLE(8192, 8192)
+            ),
+            SPA_FORMAT_VIDEO_framerate,
+            SPA_POD_CHOICE_RANGE_Fraction(
+                    &SPA_FRACTION(25, 1),
+                    &SPA_FRACTION(0, 1),
+                    &SPA_FRACTION(1000, 1)
+            )
+    );
+
+    DEBUG_SCREENCAST("screenId#%i: stream connecting %p\n", node, stream);
+
+    return fp_pw_stream_connect(
+            stream,
+            PW_DIRECTION_INPUT,
+            node,
+            PW_STREAM_FLAG_AUTOCONNECT
+            | PW_STREAM_FLAG_MAP_BUFFERS,
+            &param,
+            1
+    ) >= 0;
+}
+
+/**
+ * @param index of a screen
+ * @return TRUE on success
+ */
+static gboolean connectStream(int index) {
+    DEBUG_SCREENCAST("@@@ using screen %i\n", index);
+    if (index >= screenSpace.screenCount) {
+        DEBUG_SCREENCAST("⚠ Wrong index for screen\n", NULL);
```

Comment:
I've seen this "! in a triangle" symbol in a quite a few places in the DEBUG messages.
Looks like you got some non-ascii char in there.

Reply:
Fixed.